### ISSUE TITLE
[API Proposal]: Marshal.GetFunctionPointerForDelegate() consistent with delegate*

### ISSUE THREAD

**Issue by jonpryor** [dotnet/runtime#65853]
## Background

C#9 added [function pointers](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers).

[.NET 5 added](https://devblogs.microsoft.com/dotnet/improvements-in-native-code-interop-in-net-5-0/#unmanagedcallersonly) the [`[UnmanagedCallersOnly]`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedcallersonlyattribute?view=net-6.0) custom attribute.

Together, these allow obtaining a function pointer to a managed method and passing that function pointer to native code:

```csharp
[DllImport("user32", …)
static extern int EnumWindows(
    delegate* unmanaged<IntPtr, IntPtr, int> lpEnumFunc,
    IntPtr lParam);

[UnmanagedCallersOnly]
static int MyEnumWindowCallback (IntPtr hwnd, IntPtr lParam) {…}

void Demo() {
    delegate* unmanaged<IntPtr, IntPtr, int> enumFunc = &MyEnumWindowCallback;
    int found = EnumWindows(enumFunc, IntPtr.Zero);
}
```

.NET 1.0 supported something similar, with delegates marshaling to function pointers.  Or, [`Marshal.GetFunctionPointerForDelegate()`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getfunctionpointerfordelegate?view=net-6.0#system-runtime-interopservices-marshal-getfunctionpointerfordelegate(system-delegate)) could be used:

```csharp
[DllImport("user32", …)
static extern int EnumWindows(IntPtr lpEnumFunc, IntPtr lParam);

[UnmanagedCallersOnly]
static int MyEnumWindowCallback (IntPtr hwnd, IntPtr lParam) {…}

delegate int EnumWindowsProc(IntPtr hwnd, IntPtr lParam);

void Demo() {
    var a = Delegate.CreateDelegate(typeof(EnumWindowsProc), typeof(…), nameof(MyEnumWindowCallback));
    int found = EnumWindows(Marshal.GetFunctionPointerForDelegate(a), IntPtr.Zero);
}
```

*Note*: `[UnmanagedCallersOnly]` does not prevent looking up the method via reflection; `Delegate.CreateDelegate()` to an `[UnmanagedCallersOnly]`-attributed method *works* with .NET 6 (in that no exception is thrown, and a non-null value is returned).

### API Proposal

It would be "nice" if `Marshal.GetFunctionPointerForDelegate()` returned the same value as `delegate*` with operator `&`.  Right now, they are not the same:

```csharp
using System.Reflection;
using System.Runtime.InteropServices;

class App {
    [UnmanagedCallersOnly]
    static void n_marshaler ()
    {
    }

    public static unsafe void Main ()
    {
        delegate* unmanaged<void> d1 = &n_marshaler;
        Console.WriteLine($"n_marshaler={((IntPtr) d1).ToString ("x")}");

        var d2 = Marshal.GetFunctionPointerForDelegate(
            Delegate.CreateDelegate(typeof(Action), typeof(App), nameof(n_marshaler)));
        Console.WriteLine($"p={d2.ToString ("x")}");
    }
}
```

Currently, these return different values:

```
n_marshaler=11ae7f320
p=11a3330a4
```

### Rationale

Why does it matter?  It might not!

See also: https://github.com/dotnet/runtime/issues/32963#issuecomment-595971199

That said, the Android team has lots of existing infrastructure which is based on `System.Delegate` marshaling -- which I'm currently assuming is identical to/synonymous with `Marshal.GetFunctionPointerForDelegate()` -- and we would like to begin marking our generated "JNI marshal methods" with `[UnmanagedCallersOnly]`.  *However*, for the immediate future, *all* lookup and eventual registration of JNI marshal methods is via Reflection.  We are looking into an alternate registration approach that would avoid the need for Reflection in some circumstances, and the use of `[UnmanagedCallersOnly]` would help; see also: https://discord.com/channels/732297728826277939/732297965019988138/946331862459113483

*If* `delegate*`+`&` gave the same value as `Marshal.GetFunctionPointerForDelegate()`/`delegate` marshaling, then I think I could reasonably believe that these two different codepaths would behave consistently, and consistent behavior makes it easier for us to reason about interactions.  Otherwise, we need to wonder how the native function pointers differ…

### API Usage

-

**Comment by lambdageek** [comment]
For mono this should be _possible_.  Both codepaths end up calling `mono_marshal_get_managed_wrapper` to create a native-to-managed wrapper and return a pointer to the wrapper.  The difference is when we come in from method-to-ir `ldftn` the delegate is NULL, while on the other codepath `GetFunctionPointerForDelegate` passes the created delegate.  In principle `CreateDelegate` could note that it is being applied to an `UnmanagedCallersOnly` method, and then `GetFunctionPointerForDelegate` could inform `mono_marshal_get_managed_wrapper` about it.

**Comment by lambdageek** [comment]
Is this really an API Proposal? It seems like an implementation detail of the runtimes.

**Comment by teo-tsirpanis** [comment]
I've read somewhere in this repo that comparing managed function pointers is undefined and discouraged. Two pointers might point to the same method but have different values (the one might be a JIT stub or the method compiled at a lower tier).

**Comment by teo-tsirpanis** [comment]
Wait a minute, I don't think that the final code snippet is correct; you are not supposed to create a delegate from an `UnmanagedCallersOnly` function; these are allowed to be called only from an unmanaged function pointer.

**Comment by jkoritzinsky** [comment]
For CoreCLR, the story is a little different. CoreCLR does not generated a native-to-managed wrapper in the `UnmanagedCallersOnly` case; the native code for the method conforms to the specified native calling convention and includes the GC transition in the method.

For your example that "works", you'll end up in a bad state after the delegate call as the `UnmanagedCallersOnly` method will flip the GC mode and you'll be running managed code while the runtime thinks you're in unmanaged code. If you call `GetFunctionPointerForDelegate` on this delegate, you would end up doing the transition twice, so you'd end up in the right GC mode afterwards, but it might put the runtime into a weird state as it would transition from native to managed twice and then from managed to native twice. Additionally, if you use any custom calling conventions that have an effect on your platform in your `[UnmanagedCallersOnly]` attribute, you'd end up in a really bad state as you'd have a calling convention mismatch between the caller and callee.

To get a function pointer for an `UnmanagedCallersOnly` method through reflection, you should use [`RuntimeMethodHandle.GetFunctionPointer()`](https://docs.microsoft.com/en-us/dotnet/api/system.runtimemethodhandle.getfunctionpointer?view=net-6.0), as this method returns the actual function pointer to the method without needing to wrap in any delegates. This is the supported mechanism to get an unmanaged function pointer for an `[UnmanagedCallersOnly]` method via reflection with no overhead (and I believe this works in Mono as well).

I guess we could consider enabling `Delegate.CreateDelegate` to create a delegate like `GetDelegateForFunctionPointer` when an `[UnmanagedCallersOnly]` `MethodInfo` is passed in, but I think we'd be more likely to block that scenario as we don't enable calling an `[UnmanagedCallersOnly]` method directly in C# (you have to go through the unmanaged function pointer). Enabling this scenario falls more into the "guess what the user wants and do it" than the "tell the user when they're doing something weird and there's a clearer mechanism to do things" idea that the interop team is pushing towards.

cc: @dotnet/interop-contrib

**Comment by jonpryor** [comment]
@jkoritzinsky [wrote](https://github.com/dotnet/runtime/issues/65853#issuecomment-1050140999):

> To get a function pointer for an `UnmanagedCallersOnly` method through reflection, you should use [RuntimeMethodHandle.GetFunctionPointer()](https://docs.microsoft.com/en-us/dotnet/api/system.runtimemethodhandle.getfunctionpointer?view=net-6.0),

"Funny thing" about that: the docs say:

> The returned [IntPtr](https://docs.microsoft.com/en-us/dotnet/api/system.intptr?view=net-6.0) is meant for internal runtime usage and should not be used to call the function from native code. Use [Marshal.GetFunctionPointerForDelegate](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getfunctionpointerfordelegate?view=net-6.0) in order to get a function pointer that can be passed to native code.

i.e. the `RuntimeMethodHandle.GetFunctionPointer()` docs say (1) "don't use me", and (2) "use `Marshal.GetFunctionPointerForDelegate()`.  Which is what I was asking for? 🤔

> This is the supported mechanism to get an unmanaged function pointer for an `[UnmanagedCallersOnly]` method via reflection with no overhead (and I believe this works in Mono as well).

Should the docs be updated to say that?

**Comment by lambdageek** [comment]
> To get a function pointer for an UnmanagedCallersOnly method through reflection, you should use [RuntimeMethodHandle.GetFunctionPointer()](https://docs.microsoft.com/en-us/dotnet/api/system.runtimemethodhandle.getfunctionpointer?view=net-6.0), as this method returns the actual function pointer to the method without needing to wrap in any delegates. This is the supported mechanism to get an unmanaged function pointer for an [UnmanagedCallersOnly] method via reflection with no overhead (and I believe this works in Mono as well).

Yep. That should return the same pointer that `&method` would return in Mono, AFAICT.

**Comment by jkoritzinsky** [comment]
@jonpryor yes, we should update the documentation. That statement in the docs was from before function pointers were available in C#. At that time, there was no valid use case for the return value of the `GetFunctionPointer()` method in standard C#. Now with function pointers, the value is more useful.

**Comment by jonpryor** [comment]
Additionally, as Android is "delegates all the way down", it would significantly simplify things if `Marshal.GetFunctionPointerForDelegate()`/`delegate` P/Invoke marshaling could "special-case" `[UnmanagedCallersOnly]`-attributed methods, so that we wouldn't need to update code to use `RuntimeMethodHandle.GetFunctionPointer()` in order to get the benefits of `[UnmanagedCallersOnly]`.

The problem, from our end, is ABI and backward compatibility: all existing bindings use `delegate` marshaling everywhere.  If we have to use `RuntimeMethodHandle.GetFunctionPointer()`, then we'll need to have two codepaths, and some form of determining which codepath to use, in a way which doesn't harm our app startup times.

**Comment by jkoritzinsky** [comment]
> Additionally, as Android is "delegates all the way down", it would significantly simplify things if `Marshal.GetFunctionPointerForDelegate()`/`delegate` P/Invoke marshaling could "special-case" `[UnmanagedCallersOnly]`-attributed methods, so that we wouldn't need to update code to use `RuntimeMethodHandle.GetFunctionPointer()` in order to get the benefits of `[UnmanagedCallersOnly]`.

All of the work to special-case `[UnmanagedCallersOnly]` would be in the reflection stack (`Delegate.CreateDelegate`). Once we get to the call to `GetFunctionPointerForDelegate` or delegate marshalling it's already too late to efficiently determine if the target method is an `[UnmanagedCallersOnly]` method, and there's no good way to handle a multicast delegate created from two `UnmanagedCallersOnly` methods correctly in the interop subsystem. And I don't know how much appetite there is for changing the behavior in the reflection subsystem for this case.

> The problem, from our end, is ABI and backward compatibility: all existing bindings use `delegate` marshaling everywhere. If we have to use `RuntimeMethodHandle.GetFunctionPointer()`, then we'll need to have two codepaths, and some form of determining which codepath to use, in a way which doesn't harm our app startup times.

If you need to get a delegate for an `UnmanagedCallersOnly` method, you can call `Marshal.GetDelegateForFunctionPointer` on the return value from `RuntimeMethodHandle.GetFunctionPointer()`. As mentioned previously, the docs aren't quite up-to-date in this area. This is slightly less efficient, but if you prefer to do delegate-based interop, it should work for your scenario. The only change you would need to make is at the call-site for `Delegate.CreateDelegate`.

Basically, instead of `Delegate del = Delegate.CreateDelegate(typeof(Action), typeof(App), nameof(n_marshaler));`, you would do the following:

```csharp
MethodInfo method = typeof(App).GetMethod("n_marshaler");
Delegate del = method.IsDefined(typeof(UnmanagedCallersOnlyAttribute))
    ? Marshal.GetDelegateForFunctionPointer(typeof(Action), method.MethodHandle.GetFunctionPointer())
    : Delegate.CreateDelegate(typeof(Action), method);
```

With this snippet, you now have a delegate that is safe to call from managed code and do all the things you can do to a delegate (including calling `GetFunctionPointerForDelegate` on it whenever you want).

Would this work for the Xamarin.Android scenario?

**Comment by jonpryor** [comment]
@jkoritzinsky [wrote](https://github.com/dotnet/runtime/issues/65853#issuecomment-1050207321):

> Would this work for the Xamarin.Android scenario?

I think it would.  I'll need to consider this further.

**Comment by AaronRobinsonMSFT** [comment]
@jonpryor Based on @jkoritzinsky's comment at https://github.com/dotnet/runtime/issues/65853#issuecomment-1050140999 and with recommendation at https://github.com/dotnet/runtime/issues/65853#issuecomment-1050207321 I am inclined to close this issue. Please let us know if you think this isn't satisfactory and we should reconsider.
